web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiClient.java

```diff
@@ -0,0 +1,76 @@
+package dev.langchain4j.web.search.searchapi;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import lombok.Builder;
+import okhttp3.OkHttpClient;
+import okhttp3.ResponseBody;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.gson.FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES;
+
+class SearchApiClient {
+
+    private static final Gson GSON = new GsonBuilder()
+            .setFieldNamingPolicy(LOWER_CASE_WITH_UNDERSCORES)
+            .setPrettyPrinting()
+            .create();
+
+    private final SearchApiWebSearchApi api;
+
+    @Builder
+    SearchApiClient(String baseUrl, Duration timeout) {
+        OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder()
+                .callTimeout(timeout)
+                .connectTimeout(timeout)
+                .readTimeout(timeout)
+                .writeTimeout(timeout);
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(baseUrl)
```

Comment:
Normally your Builder class should validate If  `baseUrl` is mandatory for the API. If you're using lombok could add them in your main constructor:

```suggestion
        if (isNullOrBlank(baseUrl)) {
            throw new IllegalArgumentException("baseUrl cannot be null or empty");
        }
        Retrofit retrofit = new Retrofit.Builder()
                        .baseUrl(baseUrl)
```

Reply:
or just `.baseUrl(ensureNotBlank(baseUrl, "baseUrl"))`

Reply:
@czelabueno even this being already validated in `SearchApiWebSearchEngine` constructor which uses this builder?

Reply:
@zambrinf Yeah, what I mean is that your `SearchApiClient` should have all the constraints validated for a well communication with `SearchApi`. `SearchApiWebSearchEngine` uses `SearchApiClient` and it should have the only responsibility of passing user values ​​or setting default values. My observation is more related to putting the validation where it makes the most sense.